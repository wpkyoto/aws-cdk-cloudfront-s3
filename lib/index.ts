import * as cdk from '@aws-cdk/core';
import { OriginAccessIdentity, CloudFrontWebDistribution, LambdaEdgeEventType } from '@aws-cdk/aws-cloudfront';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import { Runtime, Version } from '@aws-cdk/aws-lambda';
import { RemovalPolicy, CfnOutput, Construct } from '@aws-cdk/core';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs'
import {createHash} from 'crypto'
import { readFileSync } from 'fs';
import { join } from 'path';


export interface CloudfrontS3Props {
  /**
   * Resource name
   */
  name: string;

  /**
   * Certification ARN for CloudFront distribution
   */
  acmCertificationARN?: string;

  /**
   * CloudFront domain names
   * If acmCertificationARN is empty, the parameter does not useed
   */
  domains?: string[]

  /**
   * Import own S3 bucket
   */
  s3Bucket?: IBucket

  /**
   * Log Lambda@edge process
   */
  debugMode?: boolean;
}

export class CloudfrontS3 extends cdk.Construct {
  /**
   * Lambda@edge function
   */
  public readonly OriginRequestHandler: NodejsFunction

  /**
   * Lambda@edge function version
   */
  public readonly OriginRequestHandlerVersion: Version

  /**
   * S3 bucket
   */
  public readonly WebsiteBucket?: IBucket

  /**
   * Origin Access Identity
   */
  public readonly DistributionOAI: OriginAccessIdentity

  /**
   * CloudFront distribution
   */
  public readonly Distribution: CloudFrontWebDistribution

  public readonly resources: Construct[] = []
  getS3Bucket(s3Bucket?: IBucket): IBucket {
    if (s3Bucket) return s3Bucket
    const WebsiteBucket = new Bucket(this, 'WebsiteBucket',{
      removalPolicy: RemovalPolicy.DESTROY
    })
    this.resources.push(WebsiteBucket)
    return WebsiteBucket
  }

  constructor(scope: cdk.Construct, id: string, props: CloudfrontS3Props) {
    super(scope, id);
    const {
      name,
      acmCertificationARN,
      domains,
      s3Bucket,
      debugMode,
    } = props;
    const EDGE_URL_HANDLER_PATH = join(__dirname, '/lambda/originRequest.ts')
  
    /**
     * Lambda Edge Functions
     */
    this.OriginRequestHandler = new NodejsFunction(this, 'OriginRequestHandler', {
      handler: 'handler',
      runtime: Runtime.NODEJS_12_X,
      functionName: `${name}URLFormatter`,
      entry: EDGE_URL_HANDLER_PATH,
      environment: {
        DEBUG: debugMode ? 'true': 'false'
      }
    })
    this.resources.push(this.OriginRequestHandler)
  
    /**
     * ZIPファイルが変更されたら新しいバージョンを発行する
     */
    this.OriginRequestHandlerVersion = this.OriginRequestHandler.addVersion(
      createHash('sha256').update(
        readFileSync(
          EDGE_URL_HANDLER_PATH,
          'utf-8'
        )
      ).digest('hex')
    )

    /**
     * S3 bucket
     */
    this.WebsiteBucket = this.getS3Bucket(s3Bucket)

    /**
     * CloudFront Distribution
     */

    this.DistributionOAI = new OriginAccessIdentity(this, 'OAIForBucket')
    this.WebsiteBucket.grantRead(this.DistributionOAI)
    this.Distribution = new CloudFrontWebDistribution(this, 'CDN', {
      aliasConfiguration: acmCertificationARN && domains ? {
          acmCertRef: acmCertificationARN,
          names: domains,
        }: undefined,
      defaultRootObject: 'index.html',
      originConfigs: [{
        s3OriginSource: {
          s3BucketSource: this.WebsiteBucket,
          originAccessIdentity: this.DistributionOAI
        },
        behaviors: [{
          isDefaultBehavior: true,
          compress: true,
          lambdaFunctionAssociations: [{
            lambdaFunction: this.OriginRequestHandlerVersion,
            eventType: LambdaEdgeEventType.ORIGIN_REQUEST,
          }]
        }]
      }]
    })
    this.Distribution.node.addDependency(this.OriginRequestHandler)
    this.resources.push(this.Distribution)
  

    new CfnOutput(this, 'CloudFrontDomain', {
      value:`https://${this.Distribution.domainName}`
    })

  }
}
