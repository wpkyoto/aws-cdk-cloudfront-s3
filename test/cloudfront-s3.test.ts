import { expect as expectCDK, haveResource, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as CloudfrontS3 from '../lib/index';
import { Bucket } from '@aws-cdk/aws-s3';

describe('Resource creation', () => {
  describe('default', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    new CloudfrontS3.CloudfrontS3(stack, 'MyTestConstruct', {
      name: 'example',
    });
    it.each([
      'AWS::IAM::Role',
      'AWS::Lambda::Function',
      'AWS::Lambda::Version',
      'AWS::S3::Bucket',
      'AWS::S3::BucketPolicy',
      'AWS::CloudFront::CloudFrontOriginAccessIdentity',
      'AWS::CloudFront::Distribution',
    ])('Should create a %p resourec', type => {
      expectCDK(stack).to(haveResource(type));
    });
    it('should match snapshot', () => {
      expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    })
  });
  describe('import s3 bucket', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    new CloudfrontS3.CloudfrontS3(stack, 'MyTestConstruct', {
      name: 'example',
      s3Bucket: Bucket.fromBucketName(stack, 'Dummy', 'dummy'),
    });
    it.each([
      'AWS::IAM::Role',
      'AWS::Lambda::Function',
      'AWS::Lambda::Version',
      'AWS::CloudFront::CloudFrontOriginAccessIdentity',
      'AWS::CloudFront::Distribution',
    ])('Should create a %p resourec', type => {
      expectCDK(stack).to(haveResource(type));
    });
    it.each(['AWS::S3::Bucket', 'AWS::S3::BucketPolicy'])(
      'Should not create a %p resourec',
      type => {
        expectCDK(stack).notTo(haveResource(type));
      }
    );
    it('should match snapshot', () => {
      expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    })
  });
  describe('with custom domain', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    new CloudfrontS3.CloudfrontS3(stack, 'MyTestConstruct', {
      name: 'example',
      domains: ['example.com'], acmCertificationARN: 'dummyarn',
    });
    it.each([
      'AWS::IAM::Role',
      'AWS::Lambda::Function',
      'AWS::Lambda::Version',
      'AWS::CloudFront::CloudFrontOriginAccessIdentity',
      'AWS::CloudFront::Distribution',
    ])('Should create a %p resourec', type => {
      expectCDK(stack).to(haveResource(type));
    });
    it.each(['AWS::S3::Bucket', 'AWS::S3::BucketPolicy'])(
      'Should not create a %p resourec',
      type => {
        expectCDK(stack).notTo(haveResource(type));
      }
    );
    it('should match snapshot', () => {
      expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    })
  })
});