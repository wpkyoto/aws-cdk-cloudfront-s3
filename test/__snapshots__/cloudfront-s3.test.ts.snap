// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource creation default should match snapshot 1`] = `
Object {
  "Outputs": Object {
    "MyTestConstructCloudFrontDomain12BA5294": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Fn::GetAtt": Array [
                "MyTestConstructCDNCFDistribution170B2D87",
                "DomainName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246ArtifactHash4CFAADFB": Object {
      "Description": "Artifact hash for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB": Object {
      "Description": "S3 bucket for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C": Object {
      "Description": "S3 key for asset version \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructCDNCFDistribution170B2D87": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerE7D62A1F",
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
        "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
      ],
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-request",
                "LambdaFunctionARN": Object {
                  "Ref": "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
                },
              },
            ],
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "MyTestConstructWebsiteBucketA07D6BA1",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestConstructOAIForBucket8661ED2F",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructOAIForBucket8661ED2F": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "MyTestConstructOriginRequestHandlerE7D62A1F": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "exampleURLFormatter",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructOriginRequestHandlerServiceRole827B3D8C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructOriginRequestHandlerE7D62A1F",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "MyTestConstructWebsiteBucketA07D6BA1": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestConstructWebsiteBucketPolicy6871CCFA": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestConstructWebsiteBucketA07D6BA1",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructOAIForBucket8661ED2F",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructWebsiteBucketA07D6BA1",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestConstructWebsiteBucketA07D6BA1",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;

exports[`Resource creation import s3 bucket should match snapshot 1`] = `
Object {
  "Outputs": Object {
    "MyTestConstructCloudFrontDomain12BA5294": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Fn::GetAtt": Array [
                "MyTestConstructCDNCFDistribution170B2D87",
                "DomainName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246ArtifactHash4CFAADFB": Object {
      "Description": "Artifact hash for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB": Object {
      "Description": "S3 bucket for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C": Object {
      "Description": "S3 key for asset version \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructCDNCFDistribution170B2D87": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerE7D62A1F",
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
        "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
      ],
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-request",
                "LambdaFunctionARN": Object {
                  "Ref": "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
                },
              },
            ],
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "dummy.s3.",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ".",
                    Object {
                      "Ref": "AWS::URLSuffix",
                    },
                  ],
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestConstructOAIForBucket8661ED2F",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructOAIForBucket8661ED2F": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "MyTestConstructOriginRequestHandlerE7D62A1F": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "exampleURLFormatter",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructOriginRequestHandlerServiceRole827B3D8C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructOriginRequestHandlerE7D62A1F",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
  },
}
`;

exports[`Resource creation with custom domain should match snapshot 1`] = `
Object {
  "Outputs": Object {
    "MyTestConstructCloudFrontDomain12BA5294": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Fn::GetAtt": Array [
                "MyTestConstructCDNCFDistribution170B2D87",
                "DomainName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246ArtifactHash4CFAADFB": Object {
      "Description": "Artifact hash for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB": Object {
      "Description": "S3 bucket for asset \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
    "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C": Object {
      "Description": "S3 key for asset version \\"25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructCDNCFDistribution170B2D87": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerE7D62A1F",
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
        "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
      ],
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.com",
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-request",
                "LambdaFunctionARN": Object {
                  "Ref": "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392",
                },
              },
            ],
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "MyTestConstructWebsiteBucketA07D6BA1",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "MyTestConstructOAIForBucket8661ED2F",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": "dummyarn",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyTestConstructOAIForBucket8661ED2F": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "MyTestConstructOriginRequestHandlerE7D62A1F": Object {
      "DependsOn": Array [
        "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3Bucket79AC7ABB",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters25a03ae01311780d8ca68ce4d0e26deed3cda9c2aedad4eca13e01840b975246S3VersionKey3690882C",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "exampleURLFormatter",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructOriginRequestHandlerServiceRole827B3D8C",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructOriginRequestHandlerServiceRole827B3D8C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "edgelambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructOriginRequestHandlerVersionffeb06218dde4c44c416e4127bfa8fd1fe3fcf33eb17be3131ac5318012afdfb84542392": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructOriginRequestHandlerE7D62A1F",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "MyTestConstructWebsiteBucketA07D6BA1": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestConstructWebsiteBucketPolicy6871CCFA": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyTestConstructWebsiteBucketA07D6BA1",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructOAIForBucket8661ED2F",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTestConstructWebsiteBucketA07D6BA1",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyTestConstructWebsiteBucketA07D6BA1",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
